packages/core/http/core-http-router-server-internal/src/versioned_router/oas_poc/generate_oas.ts

```diff
@@ -0,0 +1,234 @@
+        return !isPathParameter;
+      }
+      return true;
+    })
+    .map((shapeKey) => {
+      let shapeSchema = shape[shapeKey]!;
+      const isShapeRequired = !shapeSchema.isOptional();
+      const isPathParameter = pathParameters.includes(shapeKey);
+
+      if (!instanceofZodTypeLikeString(shapeSchema)) {
+        if (zodSupportsCoerce) {
+          if (!instanceofZodTypeCoercible(shapeSchema)) {
+            throw new Error(
+              `Input parser key: "${shapeKey}" must be ZodString, ZodNumber, ZodBoolean, ZodBigInt or ZodDate`
+            );
+          }
+        } else {
+          throw new Error(`Input parser key: "${shapeKey}" must be ZodString`);
+        }
+      }
+
+      if (instanceofZodTypeOptional(shapeSchema)) {
+        if (isPathParameter) {
+          throw new Error(`Path parameter: "${shapeKey}" must not be optional`);
+        }
+        shapeSchema = shapeSchema.unwrap();
+      }
+
+      const { description, ...openApiSchemaObject } = runtimeSchemaToJsonSchema(shapeSchema);
+
+      return {
+        name: shapeKey,
+        in: isPathParameter ? 'path' : 'query',
+        required: isPathParameter || (isRequired && isShapeRequired),
+        schema: openApiSchemaObject,
+        description,
+      };
+    });
+}
+
+function extractRequestBody(route: VersionedRouterRoute): OpenAPIV3.RequestBodyObject['content'] {
+  return route.handlers.reduce<OpenAPIV3.RequestBodyObject['content']>((acc, handler) => {
+    if (!handler.options.validate) return acc;
+    if (!handler.options.validate.request) return acc;
+    return {
+      ...acc,
+      [getVersionedContentString(handler.options.version)]: {
+        schema: runtimeSchemaToJsonSchema(handler.options.validate.request.body),
+      },
+    };
+  }, {} as OpenAPIV3.RequestBodyObject['content']);
+}
+
+function extractResponses(route: VersionedRouterRoute): OpenAPIV3.ResponsesObject {
+  return route.handlers.reduce<OpenAPIV3.ResponsesObject>((acc, handler) => {
+    if (!handler.options.validate) return acc;
+    if (!handler.options.validate.response) return acc;
+    const statusCodes = Object.keys(handler.options.validate.response);
+    for (const statusCode of statusCodes) {
+      acc[statusCode] = {
+        ...acc[statusCode],
+        description: route.options.description ?? 'No description',
+        content: {
+          ...((acc[statusCode] ?? {}) as OpenAPIV3.ResponseObject).content,
+          [getVersionedContentString(handler.options.version)]: {
+            schema: runtimeSchemaToJsonSchema(
+              handler.options.validate.response[statusCode as unknown as number].body
+            ),
+          },
+        },
+      };
+    }
+    return acc;
+  }, {});
+}
+
+function getOpenApiPathsObject(appRouter: CoreVersionedRouter): OpenAPIV3.PathsObject {
+  const routes = appRouter.getRoutes();
+  const paths: OpenAPIV3.PathsObject = {};
+  for (const route of routes) {
+    const pathParams = getPathParameters(route.path);
```

Comment:
One thing that is not mandatory in OAS but is useful to uniquely identify operations in the output is `operationId`. Ideally it would be reasonably persistent and not just a random name that differs each time the spec is generated. The Elasticsearch openAPI project generated a unique value per https://github.com/elastic/elasticsearch-openapi/pull/13 so that info might be helpful.

Reply:
That's a great idea! I think we should definitely incorporate something like this!